File: Coding_1215 1D_DP/최원준/최원준/5번.py
# https://leetcode.com/problems/decode-ways/description/

'''
1. 아이디어 :
    dp로 풀 수 있다.
    첫째 숫자가 1~9이면, 다음 dp[i+1]을 결과에 추가하고,
    첫째, 둘째 숫자가 10~26이면 dp[i+2]를 추가한다.
2. 시간복잡도 :
    O(n^2)
3. 자료구조 :
    배열
'''

from collections import deque
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = {len(s):1}

        def dfs(i):
            if i in dp:
                return dp[i]
            if s[i] == "0":
                return 0

            res = dfs(i+1)
            if (i+1<len(s) and (s[i] == "1" or s[i]=="2" and s[i+1] in "0123456")):
                res+=dfs(i+2)

            dp[i] = res
            return res

        a= dfs(0)
        print(dp)
        return a

File: Coding_231215 1D_DP/최원준/이강훈/5번.java
/*
https://leetcode.com/problems/decode-ways/description/

아이디어 ( 솔루션 보고 품 )
 ( 근데 점화식이 이해가 안댐, 왜 2개 1개 경우의 수가 합쳐지는거지..? )
 정답보고 품..

자료구조
배열

시간복잡도
O ( N )
*/

public class Solution {
    public int numDecodings(String s) {
        int[] dy = new int[s.length() + 1];

        dy[0] = 1;
        dy[1] = s.charAt(0) == '0' ? 0 : 1;

        for (int i = 2; i <= s.length(); i++) {
            int useOne = Integer.parseInt(s.substring(i - 1, i));
            if (useOne != 0) dy[i] += dy[i-1];

            int useTwo = Integer.parseInt(s.substring(i - 2, i));
            if (useTwo >= 10 && useTwo <= 26 ) dy[i] += dy[i-2];
        }

        return dy[s.length()];
    }
}
